digest/digest.go

```diff
@@ -21,31 +21,33 @@
 package digest
 
 import (
-	"hash"
 	"hash/adler32"
 
 	"github.com/m3db/m3db/ts"
+
+	"github.com/m3db/stackadler32"
 )
 
 // NewDigest creates a new digest.
 // The default 32-bit hashing algorithm is adler32.
-func NewDigest() hash.Hash32 {
-	return adler32.New()
-}
-
-// Checksum returns the 32-bit data checksum.
-func Checksum(data []byte) uint32 {
-	return adler32.Checksum(data)
+func NewDigest() stackadler32.Digest {
+	return stackadler32.NewDigest()
 }
 
-// SegmentChecksum returns the 32-bit checksum for a segment.
+// SegmentChecksum returns the 32-bit checksum for a segment
+// avoiding any allocations.
 func SegmentChecksum(segment ts.Segment) uint32 {
-	d := NewDigest()
-	if head := segment.Head; head != nil {
-		d.Write(head.Get())
+	d := stackadler32.NewDigest()
+	if segment.Head != nil {
+		d = d.Update(segment.Head.Get())
 	}
-	if tail := segment.Tail; tail != nil {
-		d.Write(tail.Get())
+	if segment.Tail != nil {
+		d = d.Update(segment.Tail.Get())
 	}
 	return d.Sum32()
 }
+
+// Checksum returns the checksum for a buffer.
+func Checksum(buf []byte) uint32 {
+	return adler32.Checksum(buf)
```

Comment:
Not really, but mainly so that only `digest` package needs to know about which implementation we're using - that way if we ever update to something other than adler32 we only need to change all references to adler32 in this one package (encapsulation)